File: tools/server_observer_cpp/src/hub_interface_wrapper.cpp
#include "hub_interface_wrapper.hpp"
#include <iostream>
#include <stdexcept>
#include <sstream>
#include <mutex>

namespace py = pybind11;

// Python initialization guard
static bool python_initialized = false;
static std::mutex python_init_mutex;
static py::scoped_interpreter* interpreter = nullptr;

json AuthDetails::to_json() const {
    return json{
        {"auth", auth},
        {"rights", rights},
        {"user_id", user_id},
        {"auth_tstamp", auth_tstamp}
    };
}

AuthDetails AuthDetails::from_json(const json& j) {
    AuthDetails details;
    details.auth = j.value("auth", "");
    details.rights = j.value("rights", "");
    details.user_id = j.value("user_id", 0);
    details.auth_tstamp = j.value("auth_tstamp", 0);
    return details;
}

HubGameProperties HubGameProperties::from_json(const json& j) {
    HubGameProperties props;
    props.game_id = j.value("game_id", 0);
    props.scenario_id = j.value("scenario_id", 0);
    props.open_slots = j.value("open_slots", 0);
    props.name = j.value("name", "");
    return props;
}

HubInterfaceWrapper::HubInterfaceWrapper(const std::string& proxy_http, const std::string& proxy_https)
    : authenticated_(false)
    , proxy_http_(proxy_http)
    , proxy_https_(proxy_https)
{
    init_python();
}

HubInterfaceWrapper::~HubInterfaceWrapper() {
    cleanup_python();
}
void HubInterfaceWrapper::init_python() {
    std::lock_guard<std::mutex> lock(python_init_mutex);

    if (!python_initialized) {
        // Initialize Python interpreter using pybind11
        interpreter = new py::scoped_interpreter();
        python_initialized = true;
    }

    py::gil_scoped_acquire acquire;

    try {
        // Set up the Python path - INSERT at beginning
        py::module_ sys = py::module_::import("sys");
        py::list path = sys.attr("path");
        
        // Insert our paths at the BEGINNING so they take priority
        path.insert(0, "/home/zdox/PycharmProjects/ConflictInterface");
        path.insert(1, "/home/zdox/PycharmProjects/ConflictInterface/.venv/lib/python3.12/site-packages");

        // Debug: Print sys.path to verify
        py::print("sys.path after setup:", path);

        // Import the HubInterface module
        python_module_ = py::module_::import("conflict_interface.interface.hub_interface");

        // Get the HubInterface class
        py::object hub_interface_class = python_module_.attr("HubInterface");

        // Create proxy dict
        py::dict proxy_dict;
        if (!proxy_http_.empty()) {
            proxy_dict["http"] = proxy_http_;
        }
        if (!proxy_https_.empty()) {
            proxy_dict["https"] = proxy_https_;
        }

        // Create HubInterface instance
        hub_interface_ = hub_interface_class(proxy_dict);

    } catch (const py::error_already_set& e) {
        throw std::runtime_error(std::string("Failed to initialize Python: ") + e.what());
    }
}

void HubInterfaceWrapper::cleanup_python() {
    // pybind11 handles cleanup automatically with RAII
    // No manual reference counting needed
}

bool HubInterfaceWrapper::login(const std::string& username, const std::string& password) {
    py::gil_scoped_acquire acquire;
    
    try {
        py::object result = hub_interface_.attr("login")(username, password);
        
        std::cout << "Logged in successfully: " << username << std::endl;
        bool success = result.cast<bool>();
        authenticated_ = true;
        return success;
    } catch (const py::error_already_set& e) {
        std::cerr << "Login failed: " << e.what() << std::endl;
        return false;
    } catch (const std::exception& e) {
        std::cerr << "Login failed: " << e.what() << std::endl;
        return false;
    }
}

void HubInterfaceWrapper::logout() {
    authenticated_ = false;
}

AuthDetails HubInterfaceWrapper::get_auth_details() const {
    py::gil_scoped_acquire acquire;
    
    AuthDetails details;
    
    try {
        // Get the auth attribute from hub_interface
        py::object auth_obj = hub_interface_.attr("auth");
        
        if (auth_obj.is_none()) {
            return details;
        }
        
        // Extract auth fields
        details.auth = auth_obj.attr("auth").cast<std::string>();
        details.rights = auth_obj.attr("rights").cast<std::string>();
        details.user_id = auth_obj.attr("user_id").cast<int>();
        details.auth_tstamp = auth_obj.attr("auth_tstamp").cast<int>();
        
    } catch (const py::error_already_set& e) {
        std::cerr << "Failed to get auth details: " << e.what() << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Failed to get auth details: " << e.what() << std::endl;
    }
    
    return details;
}

std::vector<HubGameProperties> HubInterfaceWrapper::get_my_games() {
    py::gil_scoped_acquire acquire;
    std::vector<HubGameProperties> games;
    
    try {
        py::object result = hub_interface_.attr("get_my_games")();
        if (py::isinstance<py::list>(result)) {
            json games_json = py_to_json(result);
            for (const auto& game_json : games_json) {
                games.push_back(HubGameProperties::from_json(game_json));
            }
        }
    } catch (const py::error_already_set& e) {
        std::cerr << "Failed to get my games: " << e.what() << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Failed to get my games: " << e.what() << std::endl;
    }
    
    return games;
}

std::vector<HubGameProperties> HubInterfaceWrapper::get_global_games() {
    py::gil_scoped_acquire acquire;
    std::vector<HubGameProperties> games;
    
    try {
        py::object result = hub_interface_.attr("get_global_games")();
        if (py::isinstance<py::list>(result)) {
            json games_json = py_to_json(result);
            for (const auto& game_json : games_json) {
                games.push_back(HubGameProperties::from_json(game_json));
            }
        }
    } catch (const py::error_already_set& e) {
        std::cerr << "Failed to get global games: " << e.what() << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Failed to get global games: " << e.what() << std::endl;
    }
    
    return games;
}

HubInterfaceWrapper::GameApiData HubInterfaceWrapper::join_game_as_guest(int game_id) {
    py::gil_scoped_acquire acquire;
    GameApiData data;
    data.client_version = 207;
    data.map_id = 0;
    
    try {
        // Import GameApi
        py::module_ game_api_module = py::module_::import("conflict_interface.game_api");
        py::object game_api_class = game_api_module.attr("GameApi");
        
        // Get hub_interface_.api attributes
        py::object api = hub_interface_.attr("api");
        py::object session = api.attr("session");
        py::object proxy = api.attr("proxy");
        py::object auth_details = api.attr("auth");
        
        // Create GameApi instance: GameApi(session, auth_details, game_id, proxy)
        py::object game_api = game_api_class(session, auth_details, game_id, proxy);
        
        // Call load_game_site()
        game_api.attr("load_game_site")();
        
        // Extract data from game_api
        py::object gs_addr = game_api.attr("game_server_address");
        if (!gs_addr.is_none()) {
            data.game_server_address = gs_addr.cast<std::string>();
        }
        
        py::object client_ver = game_api.attr("client_version");
        if (!client_ver.is_none()) {
            data.client_version = client_ver.cast<int>();
        }
        
        py::object map_id_obj = game_api.attr("map_id");
        if (!map_id_obj.is_none()) {
            if (py::isinstance<py::int_>(map_id_obj)) {
                data.map_id = map_id_obj.cast<int>();
            } else if (py::isinstance<py::str>(map_id_obj)) {
                data.map_id = std::stoi(map_id_obj.cast<std::string>());
            }
        }
        
        // Get updated auth
        py::object auth_obj = game_api.attr("auth");
        if (!auth_obj.is_none()) {
            data.auth.auth = auth_obj.attr("auth").cast<std::string>();
            data.auth.rights = auth_obj.attr("rights").cast<std::string>();
            data.auth.user_id = auth_obj.attr("user_id").cast<int>();
            data.auth.auth_tstamp = auth_obj.attr("auth_tstamp").cast<int>();
        }
        
        // Get session headers and cookies
        py::object ga_session = game_api.attr("session");
        py::object headers_obj = ga_session.attr("headers");
        data.headers = py_to_json(headers_obj);
        
        py::object cookies_obj = ga_session.attr("cookies");
        data.cookies = py_to_json(cookies_obj);
        
    } catch (const py::error_already_set& e) {
        std::cerr << "Failed to join game: " << e.what() << std::endl;
        throw;
    } catch (const std::exception& e) {
        std::cerr << "Failed to join game: " << e.what() << std::endl;
        throw;
    }
    
    return data;
}

json HubInterfaceWrapper::get_cookies() const {
    py::gil_scoped_acquire acquire;
    json cookies;
    
    try {
        // Get api.session.cookies
        py::object api = hub_interface_.attr("api");
        py::object session = api.attr("session");
        py::object cookies_obj = session.attr("cookies");
        cookies = py_to_json(cookies_obj);
    } catch (const py::error_already_set& e) {
        std::cerr << "Failed to get cookies: " << e.what() << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Failed to get cookies: " << e.what() << std::endl;
    }
    
    return cookies;
}

json HubInterfaceWrapper::get_headers() const {
    py::gil_scoped_acquire acquire;
    json headers;
    
    try {
        // Get api.session.headers
        py::object api = hub_interface_.attr("api");
        py::object session = api.attr("session");
        py::object headers_obj = session.attr("headers");
        headers = py_to_json(headers_obj);
    } catch (const py::error_already_set& e) {
        std::cerr << "Failed to get headers: " << e.what() << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Failed to get headers: " << e.what() << std::endl;
    }
    
    return headers;
}

json HubInterfaceWrapper::py_to_json(const py::handle& obj) const {
    if (obj.is_none()) {
        return nullptr;
    }
    
    if (py::isinstance<py::bool_>(obj)) {
        return obj.cast<bool>();
    }
    
    if (py::isinstance<py::int_>(obj)) {
        return obj.cast<long>();
    }
    
    if (py::isinstance<py::float_>(obj)) {
        return obj.cast<double>();
    }
    
    if (py::isinstance<py::str>(obj)) {
        return obj.cast<std::string>();
    }
    
    if (py::isinstance<py::list>(obj)) {
        json arr = json::array();
        py::list list = obj.cast<py::list>();
        for (const auto& item : list) {
            arr.push_back(py_to_json(item));
        }
        return arr;
    }
    
    if (py::isinstance<py::dict>(obj)) {
        json dict = json::object();
        py::dict py_dict = obj.cast<py::dict>();
        for (const auto& item : py_dict) {
            std::string key = py::str(item.first).cast<std::string>();
            dict[key] = py_to_json(item.second);
        }
        return dict;
    }
    
    // Try to convert object to dict using __dict__
    if (py::hasattr(obj, "__dict__")) {
        py::object dict = obj.attr("__dict__");
        if (py::isinstance<py::dict>(dict)) {
            return py_to_json(dict);
        }
    }
    
    return nullptr;
}

File: services/server_observer/src/hub_interface_wrapper.rs
use pyo3::prelude::*;
use pyo3::types::PyDict;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::Path;
use std::sync::Once;
use thiserror::Error;

static PY_INIT: Once = Once::new();

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct AuthDetails {
    pub auth: String,
    pub rights: String,
    pub user_id: i64,
    pub auth_tstamp: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct HubGameProperties {
    pub game_id: i32,
    pub scenario_id: i32,
    pub open_slots: i32,
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct GameApiData {
    pub game_server_address: String,
    pub client_version: i32,
    pub map_id: String,
    pub auth: AuthDetails,
    pub headers: HashMap<String, String>,
    pub cookies: HashMap<String, String>,
}

#[derive(Debug, Error)]
pub enum HubInterfaceError {
    #[error("python error: {0}")]
    Python(#[from] PyErr),
}

pub struct HubInterfaceWrapper {
    hub_interface: Py<PyAny>,
    authenticated: bool,
}

impl HubInterfaceWrapper {
    pub fn new(proxy_http: impl Into<String>, proxy_https: impl Into<String>) -> Result<Self, HubInterfaceError> {
        let proxy_http = proxy_http.into();
        let proxy_https = proxy_https.into();

        PY_INIT.call_once(Python::initialize);

        let hub_interface = Python::attach(|py| -> Result<Py<PyAny>, HubInterfaceError> {
            ensure_python_path(py)?;

            let module = py.import("conflict_interface.interface.hub_interface")?;
            let class = module.getattr("HubInterface")?;

            let proxy_dict = PyDict::new(py);
            if !proxy_http.is_empty() {
                proxy_dict.set_item("http", proxy_http)?;
            }
            if !proxy_https.is_empty() {
                proxy_dict.set_item("https", proxy_https)?;
            }

            let instance = class.call1((proxy_dict,))?;
            Ok(instance.unbind())
        })?;

        Ok(Self {
            hub_interface,
            authenticated: false,
        })
    }

    pub fn login(&mut self, username: &str, password: &str) -> Result<bool, HubInterfaceError> {
        Python::attach(|py| -> Result<bool, HubInterfaceError> {
            let hub = self.hub_interface.bind(py);
            hub.call_method1("login", (username, password))?;
            Ok(true)
        })?;
        self.authenticated = true;
        Ok(true)
    }

    #[allow(dead_code)]
    pub fn logout(&mut self) {
        self.authenticated = false;
    }

    pub fn is_authenticated(&self) -> bool {
        self.authenticated
    }

    #[allow(dead_code)]
    pub fn get_auth_details(&self) -> Result<AuthDetails, HubInterfaceError> {
        Python::attach(|py| -> Result<AuthDetails, HubInterfaceError> {
            let hub = self.hub_interface.bind(py);
            let auth_obj = hub.getattr("auth")?;
            if auth_obj.is_none() {
                return Ok(AuthDetails::default());
            }

            Ok(AuthDetails {
                auth: extract_attr_string(&auth_obj, "auth"),
                rights: extract_attr_string(&auth_obj, "rights"),
                user_id: extract_attr_i64(&auth_obj, "user_id"),
                auth_tstamp: extract_attr_i64_or_string(&auth_obj, "auth_tstamp"),
            })
        })
    }

    #[allow(dead_code)]
    pub fn get_my_games(&self) -> Result<Vec<HubGameProperties>, HubInterfaceError> {
        Python::attach(|py| -> Result<Vec<HubGameProperties>, HubInterfaceError> {
            let hub = self.hub_interface.bind(py);
            let value = hub.call_method0("get_my_games")?;
            parse_game_list(&value)
        })
    }

    pub fn get_global_games(&self) -> Result<Vec<HubGameProperties>, HubInterfaceError> {
        Python::attach(|py| -> Result<Vec<HubGameProperties>, HubInterfaceError> {
            let hub = self.hub_interface.bind(py);
            let value = hub.call_method0("get_global_games")?;
            parse_game_list(&value)
        })
    }

    pub fn join_game_as_guest(&self, game_id: i32) -> Result<GameApiData, HubInterfaceError> {
        Python::attach(|py| -> Result<GameApiData, HubInterfaceError> {
            let hub = self.hub_interface.bind(py);
            let game_api_module = py.import("conflict_interface.api.game_api")?;
            let game_api_class = game_api_module.getattr("GameApi")?;

            let api = hub.getattr("api")?;
            let session = api.getattr("session")?;
            let proxy = api.getattr("proxy")?;
            let auth_details = api.getattr("auth")?;

            let game_api = game_api_class.call1((session, auth_details, game_id, proxy))?;
            game_api.call_method0("load_game_site")?;

            let game_server_address = extract_attr_string(&game_api, "game_server_address");
            let client_version = extract_attr_i64(&game_api, "client_version") as i32;
            let map_id = extract_attr_string(&game_api, "map_id");

            let auth_obj = game_api.getattr("auth")?;
            let auth = if auth_obj.is_none() {
                AuthDetails::default()
            } else {
                AuthDetails {
                    auth: extract_attr_string(&auth_obj, "auth"),
                    rights: extract_attr_string(&auth_obj, "rights"),
                    user_id: extract_attr_i64(&auth_obj, "user_id"),
                    auth_tstamp: extract_attr_i64_or_string(&auth_obj, "auth_tstamp"),
                }
            };

            let ga_session = game_api.getattr("session")?;
            let headers = py_to_string_map(&ga_session.getattr("headers")?)?;
            let cookies = py_to_string_map(&ga_session.getattr("cookies")?)?;

            Ok(GameApiData {
                game_server_address,
                client_version: if client_version > 0 { client_version } else { 207 },
                map_id,
                auth,
                headers,
                cookies,
            })
        })
    }

    pub fn shutdown_python() {}
}

fn ensure_python_path(py: Python<'_>) -> Result<(), HubInterfaceError> {
    let sys = py.import("sys")?;
    let path = sys.getattr("path")?;

    let workspace_root_path = Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .and_then(|p| p.parent());

    let workspace_root = workspace_root_path
        .and_then(|p| p.to_str())
        .unwrap_or("/workspace")
        .to_string();

    let libs_conflict_interface = workspace_root_path
        .map(|p| p.join("libs").join("conflict_interface"));

    let libs_conflict_interface_str = libs_conflict_interface
        .as_ref()
        .and_then(|p| p.to_str())
        .unwrap_or("")
        .to_string();

    let contains_workspace = path
        .call_method1("__contains__", (workspace_root.clone(),))?
        .extract::<bool>()?;
    if !contains_workspace {
        path.call_method1("insert", (0usize, workspace_root))?;
    }

    if !libs_conflict_interface_str.is_empty() {
        let contains_libs = path
            .call_method1("__contains__", (libs_conflict_interface_str.clone(),))?
            .extract::<bool>()?;
        if !contains_libs {
            path.call_method1("insert", (0usize, libs_conflict_interface_str))?;
        }
    }

    Ok(())
}

fn parse_game_list(value: &Bound<'_, PyAny>) -> Result<Vec<HubGameProperties>, HubInterfaceError> {
    let mut games = Vec::new();
    for item in value.try_iter()? {
        let item = item?;
        games.push(HubGameProperties {
            game_id: extract_attr_i64(&item, "game_id") as i32,
            scenario_id: extract_attr_i64(&item, "scenario_id") as i32,
            open_slots: extract_attr_i64(&item, "open_slots") as i32,
            name: extract_attr_string(&item, "name"),
        });
    }
    Ok(games)
}

fn py_to_string_map(value: &Bound<'_, PyAny>) -> Result<HashMap<String, String>, HubInterfaceError> {
    if let Ok(dict) = value.cast::<PyDict>() {
        return dict_to_string_map(&dict);
    }

    if value.hasattr("get_dict")? {
        let cookie_dict = value.call_method0("get_dict")?;
        if let Ok(dict) = cookie_dict.cast::<PyDict>() {
            return dict_to_string_map(&dict);
        }
    }

    if value.hasattr("items")? {
        let items = value.call_method0("items")?;
        let mut out = HashMap::new();
        for pair in items.try_iter()? {
            let pair = pair?;
            let key = pair.get_item(0)?.extract::<String>()?;
            let val = py_value_to_string(&pair.get_item(1)?);
            out.insert(key, val);
        }
        return Ok(out);
    }

    Ok(HashMap::new())
}

fn dict_to_string_map(dict: &Bound<'_, PyDict>) -> Result<HashMap<String, String>, HubInterfaceError> {
    let mut out = HashMap::new();
    for (k, v) in dict.iter() {
        let key = k.extract::<String>()?;
        let value = py_value_to_string(&v);
        out.insert(key, value);
    }
    Ok(out)
}

fn py_value_to_string(v: &Bound<'_, PyAny>) -> String {
    if let Ok(s) = v.extract::<String>() {
        return s;
    }
    if let Ok(i) = v.extract::<i64>() {
        return i.to_string();
    }
    if let Ok(f) = v.extract::<f64>() {
        return f.to_string();
    }
    if let Ok(b) = v.extract::<bool>() {
        return b.to_string();
    }
    String::new()
}

fn extract_attr_string(obj: &Bound<'_, PyAny>, attr: &str) -> String {
    obj.getattr(attr)
        .ok()
        .and_then(|v| v.extract::<String>().ok())
        .unwrap_or_default()
}

fn extract_attr_i64(obj: &Bound<'_, PyAny>, attr: &str) -> i64 {
    obj.getattr(attr)
        .ok()
        .and_then(|v| v.extract::<i64>().ok())
        .unwrap_or_default()
}

fn extract_attr_i64_or_string(obj: &Bound<'_, PyAny>, attr: &str) -> i64 {
    if let Ok(v) = obj.getattr(attr) {
        if let Ok(num) = v.extract::<i64>() {
            return num;
        }
        if let Ok(s) = v.extract::<String>() {
            if let Ok(num) = s.parse::<i64>() {
                return num;
            }
        }
    }
    0
}
